### ISSUE TITLE
Logic Problem 1 - Code Coverage

### ISSUE THREAD

**Issue by vinceynhz** [CarterLankford/MentorExercises#15]
This is not so much an algorithm as a logic problem that happened to me, but thought of it as a nice puzzle to solve. First a few concepts:

- **Code coverage:** this refers to software that analyzes other software and looks specifically to the unit testing to make sure that you have covered all possible scenarios in your code. 

For example, if we were to have this piece of code:
```java
if ( a == 0 ){
    System.out.println(“a is 0”);
}
```
You need to make sure that your unit testing covers both cases (also known as branches, see below) for variable a: in which a is 0 and in which a is not 0. 

These possibilities are called branches because if you think of the execution of your application as a single line, when we get to these decision points we are branching out of that single line, it’s a nice analogy.

Code coverage not just checks for if sentences, but also loops, exception handling, function calling, etc. A normal coverage framework will give you a percentage of how much code you have covered with your unit testing, and while we aim at 100% some times is tricky to get there. Sometimes you need to add more test cases, some times you need to rewrite the code to make it more testable. 

With all of that in mind, here is the actual problem. We have the following piece of code:

```java
for (char currentChar = 0; currentChar < 127; currentChar++) {
    if ('0' <= currentChar && currentChar <= '9') {
        System.out.println("We have a number: " + currentChar);
    }
}
```

When we run the code analyzer, we get the following warning:

![image](https://user-images.githubusercontent.com/17203661/55815303-c1e37300-5ab5-11e9-8aac-febd4a3cc0e5.png)

Why? Is there something we can do to fix the code? I will hold on the hints for now.

**Comment by CarterLankford** [comment]
This is because you cannot have a variable entered that will both be less than or equal to zero and greater than or equal to nine.  The test function would be using a truth-table like method to validate the function and would always fail on finding a false false scenario.

**Comment by vinceynhz** [comment]
Can you add the truth table you mention? And what is the meaning on each case :)

**Comment by CarterLankford** [comment]
<h1>Truth Table</h1>
We are testing all possible outcomes of the method, both with each function individually, and finally together.

First Function | Second Function | Together
----------------|--------------------|-----------
0 <= x | x <= 9 | 0 <= x & x <= 9
T | T | T 
F | T | F
T | F | F
F | F | F

The truth table will run us through all possible options, but a value doesn't exist that will allow the False, False, False output. For the TTT we could have x = 1, the FTF x = -1, TFF x = 10.

**Comment by vinceynhz** [comment]
Uhm... The truth table is ok. But your interpretation is not correct. 

Please remember that the problem does a comparison between the **value of a variable and characters (i.e. their ASCII value); the code does not do numerical comparisons**:

| A | B | A && B |
| ---  | --- | --- |
| `'0' <= x` | `x <= '9'` | `'0' <= x && x <= '9'` |

Please take above consideration and explain what does it mean to have FFF and revisit your possible values for `x`

**Comment by CarterLankford** [comment]
I didn't keep the scope in mind when answering. Let me try again :) 

The loop is trying to compare the acii value of greater than or equal to '0' (dec 48) and less than or equal to '9' (dec 57).  
 
To validate each of the truth table functions we could use the following:
TTT: x = '0'
FTF: x = 'h'
TFF: x = '!'
FFF: since there is not a char that is both less than or equal to '0' and greater than or equal to '9', we can't test the FFF option.
